### ISSUE TITLE
[A11y] ARIA commands must have an accessible name

### ISSUE THREAD

**Issue by damienrogeonbic** [amzn/amazon-payments-magento-2-plugin#1274]
We're using Magento 2.4.7-p6 and Module version Amazon Pay Module version: 5.18.2
Following our accessibility audit, here an improvement for accessibility compliancy we've been suggested by our Magento partner.

Description: ARIA requires strict adherence to syntax. In this case, when an element has been designated as role=”link”, role=”button”, or role=”menuitem”, that element must contain certain characteristics.
Error Instance Count: 3
WCAG Level: A

Example : see capture 

<img width="780" height="101" alt="Image" src="https://github.com/user-attachments/assets/3de4513c-b2c7-4e46-914a-e2116b1d277a" />

Nothing has been overridden in our themes at the template level. We should therefore work with the publisher to make the module a11y-friendly, please.
In case it helps, here's an example of an accessible tooltip (native Magento) : see capture

<img width="906" height="431" alt="Image" src="https://github.com/user-attachments/assets/37b6ee80-456d-424e-8883-b7bc422f8078" />

**Comment by beargroupdev** [comment]
Hi @damienrogeonbic ,

Could you please provide any more detail around why the tooltip is being flagged? I see that it's related to the role of "button" that's assigned, but the statement

> when an element has been designated as role=”link”, role=”button”, or role=”menuitem”, that element must contain certain characteristics

leaves me a bit in the dark; what characteristics _should_ the element contain? The [WAVE accessibility tool](https://addons.mozilla.org/en-US/firefox/addon/wave-accessibility-tool/) does not seem to be picking up on the same. We're happy to look into removing the role altogether and correcting any errant tab indices in the name of accessibility, I'm just curious about the violation and if there's an open source tool we can use to produce the same results. Thanks!

Spencer

**Comment by jvideau** [comment]
Hi @beargroupdev ,

For an element with role="button", you must add one of the following attributes: aria-label or aria-labelledby.
In the example given by @damienrogeonbic, aria-labelledby refers to the ID of the tooltip-label block to create an accessible label.
You can also simply add an aria-label with explanatory text for accessibility.
You can use this chrome extension to test => https://chromewebstore.google.com/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd

Best regards

Jonathan

**Comment by sgabhart22** [comment]
Hi @damienrogeonbic and @jvideau ,

Thanks for bringing this up and pointing us in the right direction, the axe-devtools extension is a great addition to the accessibility toolbox. I'm attaching a small patch here which should add informative text to the tooltips via the aria-label attribute.

[gh-1274.patch.txt](https://github.com/user-attachments/files/21431447/gh-1274.patch.txt)

I will also note that there seems to be another accessibility issue with the button in sandbox mode, namely that the red 'S' that renders on the button is an image that does not have any alt text. The Magento module does not have control over this however, since this is rendered by the external Amazon Pay script. Please let me know if the patch seems to help!

Thanks,
Spencer

**Comment by damienrogeonbic** [comment]
Thanks for your patch. We'll deploy in in staging during August and on Production by end August. We already noticed tue "S" button on our staging env but no impact on the current payment process so we assumed it was only dedicate to the sandbox.

**Comment by mabaud** [comment]
Hello,

@sgabhart22 It seems you miss that template in your patch => view/frontend/web/template/payment/amazon-payment-method.html

`<div class="amazon-button-column amazon-button-column-tooltip">
                    <div class="field-tooltip toggle">
                        <span class="field-tooltip-action action-help"
                              data-bind="mageInit: {'dropdown':{'activeClass': '_active'}}"
                              data-toggle="dropdown"
                              aria-haspopup="true"
                              aria-expanded="false">
                        </span>
                        <div data-bind="i18n: 'Are you an Amazon customer? Pay now with address and payment details stored in your Amazon account.'"
                             class="field-tooltip-content"
                             data-target="dropdown"
                             aria-hidden="true">
                        </div>
                    </div>`

Could you please update your patch ?

Thanks !

**Comment by damienrogeonbic** [comment]
In case you haven't been notified by last comment from mabaud ;) @beargroupdev

**Comment by sgabhart22** [comment]
Ah, good catch @mabaud , sorry for missing that the first time around! It looks like I missed the checkout tooltips altogether; the previous patch was also missing the Express Checkout button tooltip on the shipping step. And thank you for the gentle nudge as well @damienrogeonbic 🙂

Here's an updated patch with the checkout tooltips included as well, apologies again for the inconvenience!

[gh-1274.patch.txt](https://github.com/user-attachments/files/21776514/gh-1274.patch.txt)

Please let me know if you have any other issues,
Spencer

**Comment by mdel-ca** [comment]
Hello.

We ran a WCAG audit. There are still some non-compliant criteria.


**2.5.3 (A) Label in Name**

Test: Does the accessible name of the interface components contain at a minimum the label visible
on the screen?

Problem: Tthe title and the aria-label values are not the same.
```
title="Pay using the information already stored in your Amazon account"
aria-label="Amazon Pay - Use your Amazon account"
```

Solution: Remove title attribute and update aria-label value with the visible label + extra info if needed.


**1.4.5 (AA) Images of Text**

Test: Check that no image contains text that could be formatted using CSS styling.

Problem: Images of text should be text.

Solution: Replace the image with text. 


**1.3.1 (A) Info and Relationships**

Test: Do the images have appropriate text alternatives?

Problem: Image (icon) without alt attribute.

Solution: The alt attribute is mandatory. Its value must be empty if the image is decorative.


Final HTML structure should look like the code below.

```
<div class="amazonpay-merchant-shadow-root-parent-element-for-executing-modal-script"
     role="button"
     aria-label="Amazon Pay - Use your Amazon account - Pay using the information already stored in your Amazon account">
        <img src="https://m.media-amazon.com/images/G/01/AmazonPay/Maxo/logo._CB452516594_.svg" alt="Amazon Pay">
        <img src="https://m.media-amazon.com/images/G/01/AmazonPay/Maxo/sandbox_icon._CB452516595_.svg" alt="">
        <p>Use your Amazon account</p>
</div>
```

If you need further details, please ask.
Thank you.

Marie

**Comment by damienrogeonbic** [comment]
Hi @sgabhart22 In case you haven't been notified by last comment relating to accessibility improvements.
Best,

**Comment by sgabhart22** [comment]
Hi @damienrogeonbic , I was indeed notified, but have not had a chance to dive in yet. I should be able to take a closer look today or tomorrow.

Thanks,
Spencer

**Comment by sgabhart22** [comment]
Hello @mdel-ca ,

I took a look at your latest findings, and I can see that for the first and third issues you mentioned (**2.5.3 (A) Label in Name** and **1.3.1 (A) Info and Relationships**), the Magento module is not responsible for these elements; they are rendered by the external Amazon Pay script that the module pulls in to create the button. I can speak to the SA from Amazon who we work with to see if we can get the issues in front of the team responsible for maintaining that script.

As for the second issue (**1.4.5 (AA) Images of Text**), I can't seem to find where this is being flagged in a vanilla installation. Could you please provide some more detail as to what page(s) or Amazon Pay elements are triggering this?

Thanks,
Spencer

**Comment by mdel-ca** [comment]
Hello @sgabhart22 

Thank you for your prompt reply.

About the "1.4.5 (AA) Images of Text" issue: there is a sentence "Use your Amazon account" on screen. When I inspect the element, it turns out this is a SVG image called within an `<img>` tag. It should be a text within a `<p>` tag.

Here is the image url: https://static-na.payments-amazon.com/assets/maxo/microtext/en_US.svg

I guess it comes from Amazon external script as well.

It would be great if you could share our feedback with the Amazon team. Thank you!

Marie

**Comment by sgabhart22** [comment]
You're welcome @mdel-ca ! And thanks for the additional details on the "Images of Text" issue, I was having trouble pinpointing that with the extension I was using (or the free version of it, at least).

I have passed all 3 outstanding issues on to my primary contact at Amazon Pay, and he's advised that he will get them over to the team responsible for the script that does this rendering. I'll keep you updated with any news that comes back, and thanks again for providing more clarity around the accessibility problems!

Spencer

**Comment by damienrogeonbic** [comment]
Hi,
Any updates on this request ?

**Comment by damienrogeonbic** [comment]
Hi,
Do you have more visibility on this topic ? Thanks

**Comment by moniquemessman** [comment]
@damienrogeonbic I have followed up with our contact Amazon Pay and will get back to you once I hear from him.

**Comment by shangamesh** [comment]
@damienrogeonbic 
From Amazon pay side, we made certain button changes related to accessibility. Can you conform whether any issue related to amazon pay button is observed

**Comment by damienrogeonbic** [comment]
Thanks for the update. We'll update the module soon and let you know.

**Comment by mcastan-cleverage** [comment]
@sgabhart22 @shangamesh @moniquemessman @beargroupdev @damienrogeonbic 

Hi, we recently updated the module to the latest version (**5.18.3**), but we don't see any changes for the "**Use your Amazon account**" image. We still have an `<img>` tag with an `aria-label` attribute, but without an `alt` attribute:
```
<picture class="amazonpay-button-microtext"><img src="https://static-na.payments-amazon.com/assets/maxo/microtext/en_US.svg" aria-label="Use your Amazon Pay Sandbox test account"></picture>
```

As a reminder, [we recommended](https://github.com/amzn/amazon-payments-magento-2-plugin/issues/1274#issuecomment-3206902181) the following HTML structure:
```
<div class="amazonpay-merchant-shadow-root-parent-element-for-executing-modal-script"
     role="button"
     aria-label="Amazon Pay - Use your Amazon account - Pay using the information already stored in your Amazon account">
        <img src="https://m.media-amazon.com/images/G/01/AmazonPay/Maxo/logo._CB452516594_.svg" alt="Amazon Pay">
        <img src="https://m.media-amazon.com/images/G/01/AmazonPay/Maxo/sandbox_icon._CB452516595_.svg" alt="">
        <p>Use your Amazon account</p>
</div>
```

Is there anything else to do besides updating the module?

Best regards.

**Comment by damienrogeonbic** [comment]
Hi,

Can we have a status on this , thanks

**Comment by shangamesh** [comment]
We have asked internal team to take a look & fix.

However to this issue wont impact anything related to transaction or payment for buyer

**Comment by damienrogeonbic** [comment]
thanks. We know it's only related to accessibility compliance. But we're working on compliancy on our website so it's important for us to get the visibility of the deadline on your end.

thanks in advance

**Comment by dimadidr** [comment]
Hi @damienrogeonbic,

While the team is working on an update to the direct script, I've prepared a workaround on the Magento side that could help you achieve your needs. Could you please apply this patch.

`git apply ASD-1280.patch`

[ASD-1280.patch](https://github.com/user-attachments/files/23480405/ASD-1280.patch)

Thank you,
Dima

**Comment by damienrogeonbic** [comment]
Hi @dimadidr 
Thanks for the follow up. Do you have more visibility for the module update ? If it's soon, we can wait a bit, otherwise we'll apply the patch.

Thanks.

**Comment by dimadidr** [comment]
Hi @damienrogeonbic,

Thanks for checking in. Unfortunately, I don't have any visibility on the module update at the moment.

Best,
Dima

**Comment by mdel-ca** [comment]
Hello.

We applied the patch but there are still 2 issues regarding accessibility:
 - The image "Use your Amazon account" should be replaced by a real text mention (html tag `<p>` instead of `<img>`).
 - The role="button" element hasn't been fixed : the alt attribute is still present (it should be removed) and the value of the aria-label attribute is not relevant (it should be "Amazon Pay - Use your Amazon account").

Please, let us know as soon as this fix is ​​available.
Best regards.

Marie

**Comment by dimadidr** [comment]
Hi @mdel-ca,

Thank you for the information. I have updated the patch. Could you please apply this patch.

`git apply ASD-1280_1.patch`

[ASD-1280_1.patch](https://github.com/user-attachments/files/24062407/ASD-1280_1.patch)

Thank you,
Dima
